src/environment.py

```diff
@@ -0,0 +1,298 @@
+        "MM_SAMLSETTINGS_FIRSTNAMEATTRIBUTE": "fullname",
+        "MM_SAMLSETTINGS_LASTNAMEATTRIBUTE": "",
+        "MM_SAMLSETTINGS_IDPCERTIFICATEFILE": SAML_IDP_CRT,
+        # If not set, we have to install xmlsec1, and Mattermost forks on every login(!).
+        "MM_EXPERIMENTALSETTINGS_USENEWSAMLLIBRARY": use_experimental_saml_library,
+    }
+
+
+def _env_for_smtp(config: dict) -> dict:
+    """Return settings for an outgoing SMTP relay.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        dict of settings for an outgoing SMTP relay.
+    """
+    if any(
+        (
+            setting not in config
+            for setting in (
+                "smtp_host",
+                "smtp_connection_security",
+                "smtp_from_address",
+                "smtp_reply_to_address",
+                "smtp_password",
+                "smtp_user",
+                "smtp_port",
+            )
+        )
+    ):
+        return {}
+
+    if not config["smtp_host"]:
+        return {}
+
+    enable_smtp_auth = "false"
+    if config["smtp_user"] and config["smtp_password"]:
+        enable_smtp_auth = "true"
+
+    # https://github.com/mattermost/mattermost-server/blob/master/model/config.go#L1532
+    return {
+        "MM_EMAILSETTINGS_CONNECTIONSECURITY": config["smtp_connection_security"],
+        "MM_EMAILSETTINGS_ENABLESMTPAUTH": enable_smtp_auth,
+        "MM_EMAILSETTINGS_FEEDBACKEMAIL": config["smtp_from_address"],
+        "MM_EMAILSETTINGS_REPLYTOADDRESS": config["smtp_reply_to_address"],
+        "MM_EMAILSETTINGS_SMTPPASSWORD": config["smtp_password"],
+        "MM_EMAILSETTINGS_SMTPPORT": config["smtp_port"],
+        "MM_EMAILSETTINGS_SMTPSERVER": config["smtp_host"],
+        "MM_EMAILSETTINGS_SMTPUSERNAME": config["smtp_user"],
+    }
+
+
+def missing_config_settings(config: dict) -> list:
+    """Return a list of settings required to satisfy configuration dependencies.
+
+    Args:
+        config: dict of the charm's configuration
+
+    Returns:
+        a list of settings required to satisfy configuration dependencies.
+    """
+    missing = [setting for setting in REQUIRED_SETTINGS if not config.get(setting)]
+
+    if config.get("clustering") and not config.get("license"):
+        missing.append("license")
+
+    if config.get("mattermost_image_username") and not config.get("mattermost_image_password"):
+        missing.append("mattermost_image_password")
+
+    if config.get("performance_monitoring_enabled") and not config.get("license"):
+        missing.append("license")
+
+    if config.get("s3_enabled"):
+        missing += [setting for setting in REQUIRED_S3_SETTINGS if not config.get(setting)]
+
+    if config.get("s3_server_side_encryption") and not config.get("license"):
+        missing.append("license")
+
+    if config.get("sso") == "true":
+        missing += [setting for setting in REQUIRED_SSO_SETTINGS if not config.get(setting)]
+
+    return sorted(missing)
```

Comment:
Something to consider, make this function a generator that yields all the missing settings and get the caller to pass it to `sorted`. That way you don't have to keep appending to a list or combining lists which is a bit inefficient

Reply:
Changed to generator